### PR TITLE
[feature-wip](unique-key-merge-on-write) fix thread safe issue in BetaRowsetWriter

### PR DESCRIPTION
**Pull request by zhannngchen**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/olap/rowset/beta_rowset_writer.h [hunk 1]
```diff
@@ -69,6 +69,7 @@ class BetaRowsetWriter : public RowsetWriter {
     RowsetTypePB type() const override { return RowsetTypePB::BETA_ROWSET; }
 
     Status get_segment_num_rows(std::vector<uint32_t>* segment_num_rows) const override {
+        std::shared_lock l(_lock);
```
Comment (liaoxin01) [inline be/src/olap/rowset/beta_rowset_writer.h:72]: use lock_guard or change `_lock` to shared_mutex?

File: be/src/olap/rowset/beta_rowset_writer.h [hunk 2]
```diff
@@ -93,7 +94,6 @@ class BetaRowsetWriter : public RowsetWriter {
     /// Because we want to flush memtables in parallel.
     /// In other processes, such as merger or schema change, we will use this unified writer for data writing.
     std::unique_ptr<segment_v2::SegmentWriter> _segment_writer;
-    mutable SpinLock _lock; // lock to protect _wblocks.
```
Comment (liaoxin01) [inline be/src/olap/rowset/beta_rowset_writer.h:96]: SpinLock is used in `_create_segment_writer`. Can we delete it here？

File: be/src/olap/rowset/beta_rowset_writer.h (hunk 1 above)
Comment (zhannngchen) [reply be/src/olap/rowset/beta_rowset_writer.h]: fixed

**Review by liaoxin01 — DISMISSED** [review]
LGTM

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 3]
```diff
@@ -346,7 +345,13 @@ Status BetaRowsetWriter::_flush_segment_writer(std::unique_ptr<segment_v2::Segme
     DCHECK_LE(min_key.compare(max_key), 0);
     key_bounds.set_min_key(min_key.to_string());
     key_bounds.set_max_key(max_key.to_string());
-    _segments_encoded_key_bounds.emplace_back(key_bounds);
+    {
+        std::lock_guard<SpinLock> l(_lock);
+        _segment_num_rows.reserve(_num_segment);
```
Comment (liaoxin01) [inline be/src/olap/rowset/beta_rowset_writer.cpp:350]: should use `resize()` here?

Comment (zhannngchen) [reply be/src/olap/rowset/beta_rowset_writer.cpp]: emm... right
I want to use `resize(segment_id)` at first, but resize may shrink the vector size, so changed to `reserve(segment_id)`, but ignored the fact that `reserve` won't change size of vector.

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**